### ISSUE TITLE
Behavior difference between "a is not X" and "!(a is X)" patterns

### ISSUE THREAD

**Issue by BhaaLseN** [dotnet/roslyn#62078]
**Version Used**: Visual Studio 2022 v17.2.4

**Steps to Reproduce**:

1. Have a `!(variable is ClassName)` pattern in a context where `ClassName` is also a property name.
2. Invoke the CodeFix for `IDE0083` to convert this to an `is not` pattern.
3. Compilation fails after this.

See [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAOgBlYBHCgUQA8AHAJwFMBnLqAexhcA3AQLoAzGUwkAwiQDeBEspLsoANwgAXDmWykqEMMAQRm7brwEKSAcw5ahJLg6cBfJSolkALCQCyABQAlAqeKirmnDz8MCQcrNFWcQC8JDAArgA2WSL4ERFQAGYkgQCEgQkWMdZQXCSGxqZRlrHBweEFEeQAnIEARNgg6XxaJBAkWUYmE1VJsf3BeV3KAPSr8WxsfGyyAMpERJg9wzIQMDCjJFBgLFlQAMZQWlkAniQPAuocbGNaryxdABychEIEkLR8Eggyg0GD0Fo1QSMRKtARAzoRdYQgAWdRIAl0LEeAGt6hkWLjdKDVNtAb9XnASGxtDifrjzlSIQCOJiVOtiqU5mi4vjLmNGjNEckOvkVtivNg+v1MBBhuLxpNpqZ4qikYtlitBZU9clrvUNZLmqbYiQAPqylYFXoDTCgEZjCZTJqzG0CA185TG4VI80evTUOgo6rJLjUbVmP0wR1OxXK8Tqq5ehO6mMLJadDz4NxAA===) snippet:
```c#
using System;
using System.Linq.Expressions;

public class C {
    private string LambdaExpression { get; set; }
    public void M() {
        Expression expression = null;
        if (!(expression is LambdaExpression))
            Console.WriteLine("1: not a lambda expression");
        // error CS0029: Cannot implicitly convert type 'string' to 'System.Linq.Expressions.Expression'
        // this one picks up the string property, rather than the type
        //if (expression is not LambdaExpression)
        //    Console.WriteLine("2a: not a lambda expression");
        if (expression is not LambdaExpression _)
             Console.WriteLine("2b: not a lambda expression");
        if (expression is not System.Linq.Expressions.LambdaExpression)
            Console.WriteLine("3: not a lambda expression");
    }
}
```

**Expected Behavior**:
Either:
1. Code still compiles afterwards, or
2. The CodeFix for `IDE0083` does not convert the expression.

**Actual Behavior**:
Code fails to compile after invoking the CodeFix, because it attempts to match against the property name (or its value?)
I'm not sure what it is even _trying_ to do there, because it can't possibly be parsed as constant pattern; and it doesn't match as a type pattern (which would be the expected thing).
It _does_ work if I add a variable name or discard at the end (which turns it into a declaration pattern).

**Comment by jcouv** [comment]
This is expected from the language/compiler side and results from backwards compatibility concerns with `is` expressions which predate patterns.

From the [patterns doc](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.0/pattern-matching.md#pattern-matching):
"Pattern matching introduces a new context where the expression-type ambiguity arises. Previously the right-hand-side of an `is` operator was a type. Now it can be a type or expression, and if it is a type it may be followed by an identifier."

In other words, an is-expression previously would only bind to a type (and this has to be maintained for compatibility reasons), but an is-pattern can bind to a type or a (constant) expression.

I'm moving this issue to the IDE. Let's see if we can avoid offering the fix in this situation.

**Comment by BhaaLseN** [comment]
I'm ok with either. It was just a bit weird because I ran the code cleanup (specific rules added to profile 2, just to apply the IsNot preferences) on my solution with 10 projects after upgrading the language version to C#10.0 and suddenly it wouldn't compile.

In case it matters, the property was in a base class, so it wasn't immediately visible when looking at the code.

**Comment by RikkiGibson** [comment]
It would be useful to know if the issue occurs when the property is in the same class. If it happens to do something "successful" in that situation, we'd want to imitate that behavior when the property is on the base type.

**Comment by RikkiGibson** [comment]
I was referring to the "before refactoring" state, i.e. if the property is in the same class, is the refactoring offered and if it is used does it result in a similar error.

**Comment by BhaaLseN** [comment]
Oh, sorry, I misunderstood that.

Yes, I just tried this on a test project, I get the refactoring "Use pattern matching" für `IDE0083` in both cases (property in a base class, property in the same class). Both cause compilation to fail.

**Comment by CyrusNajmabadi** [comment]
Dupe of https://github.com/dotnet/roslyn/issues/72370
